hal/src/thumbv6m/sercom/v2/uart/pad.rs

```diff
@@ -0,0 +1,452 @@
+/// use atsamd_hal::sercom::v2::{Sercom0, uart};
+/// use atsamd_hal::typelevel::NoneT;
+///
+/// pub type Pads = uart::PadsFromIds<Sercom0, PA08, PA09>;
+///
+/// pub fn create_pads() -> Pads {
+///     let peripherals = Peripherals::take().unwrap();
+///     let pins = Pins::new(peripherals.PORT);
+///     uart::Pads::default().rx(pins.pa09).tx(pins.pa08)
+/// }
+/// ```
+///
+/// [`Pin`]: crate::gpio::v2::Pin
+/// [`PinId`]: crate::gpio::v2::PinId
+/// [`OptionalPinId`]: crate::gpio::v2::OptionalPinId
+
+#[cfg(feature = "samd21")]
+pub type PadsFromIds<S, RX = NoneT, TX = NoneT, RTS = NoneT, CTS = NoneT> = Pads<
+    S,
+    <RX as GetOptionalPad<S>>::Pad,
+    <TX as GetOptionalPad<S>>::Pad,
+    <RTS as GetOptionalPad<S>>::Pad,
+    <CTS as GetOptionalPad<S>>::Pad,
+>;
+
+//=============================================================================
+// PadSet
+//=============================================================================
+
+/// Type-level function to recover the [`OptionalPad`] types from a generic set
+/// of [`Pads`]
+///
+/// This trait is used as an interface between the [`Pads`] type and other
+/// types in this module. It acts as a [type-level function], returning the
+/// corresponding [`Sercom`] and [`OptionalPad`] types. It serves to cut down on
+/// the total number of type parameters needed in the [`Config`] struct. The
+/// [`Config`] struct doesn't need access to the [`Pin`]s directly.  Rather, it
+/// only needs to apply the [`SomePad`] trait bound when a `Pin` is required.
+/// The [`PadSet`] trait allows each [`Config`] struct to store an instance of
+/// [`Pads`] without itself being generic over all six type parameters of the
+/// [`Pads`] type.
+///
+/// [`Pin`]: crate::gpio::v2::Pin
+/// [type-level function]: crate::typelevel#type-level-functions
+/// [`Config`]: super::Config
+pub trait PadSet: Sealed {
+    type Sercom: Sercom;
+    type Rx: OptionalPad;
+    type Tx: OptionalPad;
+    type Rts: OptionalPad;
+    type Cts: OptionalPad;
+}
+
+impl<S, RX, TX, RTS, CTS> Sealed for Pads<S, RX, TX, RTS, CTS>
+where
+    S: Sercom,
+    RX: OptionalPad,
+    TX: OptionalPad,
+    RTS: OptionalPad,
+    CTS: OptionalPad,
+{
+}
+
+impl<S, RX, TX, RTS, CTS> PadSet for Pads<S, RX, TX, RTS, CTS>
+where
+    S: Sercom,
+    RX: OptionalPad,
+    TX: OptionalPad,
+    RTS: OptionalPad,
+    CTS: OptionalPad,
+{
+    type Sercom = S;
+    type Rx = RX;
+    type Tx = TX;
+    type Rts = RTS;
+    type Cts = CTS;
+}
+
+//=============================================================================
+// ValidPads
+//=============================================================================
+
+/// Marker trait for valid sets of [`Pads`]
+///
+/// This trait labels sets of [`Pads`] that satisfy the [`RxpoTxpo`] trait. It
+/// guarantees to the [`Config`] struct that this set of `Pads` can be
+/// configured through that trait.
+///
+/// [`Config`]: super::Config
+pub trait ValidPads: PadSet + RxpoTxpo {}
```

Comment:
I would get rid of the `GetCapability` trait and add `Capability` as an associated type on `ValidPads`. I think that makes sense, because a `PadSet` can't be valid unless it has an associated `Capability` and satisfies `RxpoTxpo`.